### ISSUE TITLE
Blank Entries and Advancing Terms when Installing Snapshot

### ISSUE THREAD

**Issue by fredfortier** [railgun-rs/actix-raft#128]
I've implement a store backend based on `MemStore`. Things like installing in `finalize_snapshot_installation` installation aren't instantaneous anymore which seems to cause problems. 

During the `compaction` integration test, I’m getting repeated calls to the finalize installation fn resulting in multiple append log blank entries each advancing the terms. What’s happening is that node 1’s, install snapshot RPC is long running. But I’d expect the leader to consider that node 1 is catching up until the rpc call finishes. Is this the intent and I'm doing something wrong? If not, any guidance around nodes catching up would be appreciated.
```
Jul 02 00:04:44.399  INFO finalize_snapshot_installation{index=17 term=1 delete_through=None id=""}: ddx_core::raft::store: enter
Jul 02 00:04:44.399  INFO finalize_snapshot_installation{index=17 term=1 delete_through=None id=""}: ddx_core::raft::store: exit
Jul 02 00:04:44.414 ERROR async_raft::replication: timeout while sending InstallSnapshot RPC to target error=deadline has elapsed
Jul 02 00:04:44.414  INFO Server::run{addr=0.0.0.0:11001}: warp::server: enter
Jul 02 00:04:44.415  INFO Server::run{addr=0.0.0.0:11001}: warp::server: exit
Jul 02 00:04:44.465 ERROR async_raft::replication: timeout while sending InstallSnapshot RPC to target error=deadline has elapsed
```
To avoid race conditions with the multiple calls, I'm tracking the snapshot installation status in the store so `finalize_snapshot_installation` returns `Ok(())` if the snapshot installation is already in progress. Is this right approach, or should I raise an exception, or something else? The heartbeat timeout is at the default 50ms. I could increase it but it does not make sense to increase the heartbeat globally when snapshot installation is fundamentally more expensive than other roc’s. Any help would be appreciated.

Based on the documentation shouldn't `.add_non_voter(0, 1)` trigger the snapshot installation rather than `.change_membership(0, hashset![0, 1])`? I'd like to get node 1 to catch up before changing the membership.

**Comment by fredfortier** [comment]
Looking at `handle_needs_snapshot` it seems like the intent is correct:

> Check if snapshot creation is already in progress. If so, we spawn a task to await its completion (or cancellation), and respond to the replication stream. The repl stream will wait for the completion and will then send anothe request to fetch the finished snapshot. Else we just drop any other state and continue. Leaders never enter `Streaming` state.

What should node 1 do, other than implementing the long running logic in `finalize_snapshot_installation` to stay in the `SnapshotState::Snapshotting` state?

**Comment by fredfortier** [comment]
It works now for all intents and purposes. I'm still confused by all the `finalize_snapshot_installation` timeouts but I understand it's stopping at replication and not actually penetrating into my store's implementation.

I also still do not understand why `.add_non_voter(0, 1)` does not cause the snapshot to synchronize before initiating the change in membership.

**Comment by drmingdrmer** [comment]
> I also still do not understand why `.add_non_voter(0, 1)` does not cause the snapshot to synchronize before initiating the change in membership.

`change_membership` is redundant for this test. Removing it does not affect the snapshot testing.
But another assertion `assert_stable_cluster` requires all nodes to be voters to pass:

https://github.com/async-raft/async-raft/blob/bcc246ccf8fa2858baf8e07bf760ee6db4d85cf6/async-raft/tests/fixtures/mod.rs#L293-L297

**Comment by fredfortier** [comment]
What I'm experiencing is that calling `.add_non_voter(0, 1)` does not trigger the snapshot, only `change_membership` does. I though it was redundant as well because the doc states that `add_non_voter` should sync the node and suggest waiting for the node to catch up before calling `change_membership`. Let me know if you're experiencing something different. I can try to better isolate this if needed.

**Comment by drmingdrmer** [comment]
I am not sure if `async-raft-v0.6.1` behave differently.
I've added several fixes upon `async-raft-v0.6.1` and with my fork it does not need `change_membership`. This is my latest test:

https://github.com/datafuse-extras/async-raft/blob/53e696872c15b4dbd56316eabae42b962f0197be/async-raft/tests/compaction.rs#L78-L92

I shouldn't have changed anything  that affect snapshotting, I hope `:)`

**Comment by fredfortier** [comment]
Yes, you're right. I'm using your fork and it works perfectly now. I'll go back to test with 0.6.1 as it's possible I was doing something incorrectly at the time of opening the issue. Any plans to submit a PR for your fork? I'm happy to help reviewing it and contributing if you need anything.
